File: android/app/src/main/kotlin/pt/ualg/miaugenda/data/local/TokenManager.kt
package pt.ualg.miaugenda.data.local

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

class TokenManager(context: Context) {
    
    private val sharedPreferences: SharedPreferences = try {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        EncryptedSharedPreferences.create(
            context,
            "miaugenda_secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    } catch (e: Exception) {
        // Fallback para SharedPreferences normal se EncryptedSharedPreferences falhar
        context.getSharedPreferences("miaugenda_prefs_fallback", Context.MODE_PRIVATE)
    }

    companion object {
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
        private const val KEY_USER_ID = "user_id"
        private const val KEY_USER_NAME = "user_name"
        private const val KEY_USER_EMAIL = "user_email"
        private const val KEY_USER_ROLE = "user_role"
        private const val KEY_USER_CATEGORY = "user_category"
        private const val KEY_EMPLOYEE_NUMBER = "employee_number"
    }

    fun saveTokens(
        accessToken: String,
        refreshToken: String,
        userId: Int,
        userName: String,
        userEmail: String,
        userRole: String,
        userCategory: String,
        employeeNumber: String
    ) {
        sharedPreferences.edit().apply {
            putString(KEY_ACCESS_TOKEN, accessToken)
            putString(KEY_REFRESH_TOKEN, refreshToken)
            putInt(KEY_USER_ID, userId)
            putString(KEY_USER_NAME, userName)
            putString(KEY_USER_EMAIL, userEmail)
            putString(KEY_USER_ROLE, userRole)
            putString(KEY_USER_CATEGORY, userCategory)
            putString(KEY_EMPLOYEE_NUMBER, employeeNumber)
            apply()
        }
    }

    fun getAccessToken(): String? {
        return sharedPreferences.getString(KEY_ACCESS_TOKEN, null)
    }

    fun getRefreshToken(): String? {
        return sharedPreferences.getString(KEY_REFRESH_TOKEN, null)
    }

    fun getUserId(): Int {
        return sharedPreferences.getInt(KEY_USER_ID, -1)
    }

    fun getUserName(): String? {
        return sharedPreferences.getString(KEY_USER_NAME, null)
    }

    fun getUserEmail(): String? {
        return sharedPreferences.getString(KEY_USER_EMAIL, null)
    }

    fun getUserRole(): String? {
        return sharedPreferences.getString(KEY_USER_ROLE, null)
    }

    fun getUserCategory(): String? {
        return sharedPreferences.getString(KEY_USER_CATEGORY, null)
    }

    fun getEmployeeNumber(): String? {
        return sharedPreferences.getString(KEY_EMPLOYEE_NUMBER, null)
    }

    fun isLoggedIn(): Boolean {
        return getAccessToken() != null
    }

    fun clearTokens() {
        sharedPreferences.edit().clear().apply()
    }

    fun updateAccessToken(accessToken: String) {
        sharedPreferences.edit().putString(KEY_ACCESS_TOKEN, accessToken).apply()
    }

    fun updateUserData(
        name: String,
        contact: String?,
        category: String
    ) {
        sharedPreferences.edit().apply {
            putString(KEY_USER_NAME, name)
            putString(KEY_USER_CATEGORY, category)
            apply()
        }
    }
}

File: desktop/src/storage/tokenManager.ts
import type { User } from "../models/user";

const KEY_ACCESS_TOKEN = "access_token";
const KEY_REFRESH_TOKEN = "refresh_token";
const KEY_USER_ID = "user_id";
const KEY_USER_NAME = "user_name";
const KEY_USER_EMAIL = "user_email";
const KEY_USER_ROLE = "user_role";
const KEY_USER_CATEGORY = "user_category";
const KEY_EMPLOYEE_NUMBER = "employee_number";

export const tokenManager = {
  saveTokens(accessToken: string, refreshToken: string, user: User): void {
    localStorage.setItem(KEY_ACCESS_TOKEN, accessToken);
    localStorage.setItem(KEY_REFRESH_TOKEN, refreshToken);
    localStorage.setItem(KEY_USER_ID, String(user.id));
    localStorage.setItem(KEY_USER_NAME, user.name);
    localStorage.setItem(KEY_USER_EMAIL, user.email);
    localStorage.setItem(KEY_USER_ROLE, user.role);
    localStorage.setItem(KEY_USER_CATEGORY, user.category);
    localStorage.setItem(KEY_EMPLOYEE_NUMBER, user.employeeNumber);
  },

  getAccessToken(): string | null {
    return localStorage.getItem(KEY_ACCESS_TOKEN);
  },

  getRefreshToken(): string | null {
    return localStorage.getItem(KEY_REFRESH_TOKEN);
  },

  getUserId(): number {
    const value = localStorage.getItem(KEY_USER_ID);
    return value ? Number(value) : -1;
  },

  getUserName(): string | null {
    return localStorage.getItem(KEY_USER_NAME);
  },

  getUserEmail(): string | null {
    return localStorage.getItem(KEY_USER_EMAIL);
  },

  getUserRole(): string | null {
    return localStorage.getItem(KEY_USER_ROLE);
  },

  getUserCategory(): string | null {
    return localStorage.getItem(KEY_USER_CATEGORY);
  },

  getEmployeeNumber(): string | null {
    return localStorage.getItem(KEY_EMPLOYEE_NUMBER);
  },

  isLoggedIn(): boolean {
    return this.getAccessToken() !== null;
  },

  clearTokens(): void {
    localStorage.removeItem(KEY_ACCESS_TOKEN);
    localStorage.removeItem(KEY_REFRESH_TOKEN);
    localStorage.removeItem(KEY_USER_ID);
    localStorage.removeItem(KEY_USER_NAME);
    localStorage.removeItem(KEY_USER_EMAIL);
    localStorage.removeItem(KEY_USER_ROLE);
    localStorage.removeItem(KEY_USER_CATEGORY);
    localStorage.removeItem(KEY_EMPLOYEE_NUMBER);
  },

  updateAccessToken(accessToken: string): void {
    localStorage.setItem(KEY_ACCESS_TOKEN, accessToken);
  },

  updateTokens(accessToken: string, refreshToken: string): void {
    localStorage.setItem(KEY_ACCESS_TOKEN, accessToken);
    localStorage.setItem(KEY_REFRESH_TOKEN, refreshToken);
  },

  updateUserData(name: string, category: string): void {
    localStorage.setItem(KEY_USER_NAME, name);
    localStorage.setItem(KEY_USER_CATEGORY, category);
  }
};